parsing/parchmint-parser.js

```diff
@@ -116,46 +126,87 @@ class ParchmintParser {
+     * @param {object} jsonObj  A parsed JSON object representing the Parchmint
+     *                          file.
      */
-    initParchKey(parchKeyObj, jsonObj) {
-        parchKeyObj.name = jsonObj['name'];
-        parchKeyObj.id = jsonObj['id'];
+    parseLayers(jsonObj) {
+        // First make sure that neither components nor connections has extra/invalid layers
+        for (let layerID of this.components.keys()) {
+            let found = false;
+            for (let l of jsonObj['layers']) {
+                if (l.id === layerID) {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found) {
+                this.valid = false;
+                console.log('Parser: The Components list contains an invalid Layer (' + layerID + ').');
+            }
+        }
+
+        for (let layerID of this.connections.keys()) {
+            let found = false;
+            for (let l of jsonObj['layers']) {
+                if (l.id === layerID) {
+                    found = true;
+                    break;
+                }
+            }
+
+            if (!found) {
+                this.valid = false;
+                console.log('Parser: The Connections list contains an invalid Layer (' + layerID + ').');
+            }
+        }
+
+        // Now we can move on to the actual parsing.
+        jsonObj['layers'].forEach((value, index) => {
+            if (this.isUniqueID(value['id'])) {
+                this.layers.push(this.getParsedLayer(value));
+            } else {
+                this.valid = false;
+                console.log('Parser: Duplicate IDs (' + layerID + ') found in the "layers" key. Skipping Layer' +
+                        ' with name "' + value['name'] + '" at index ' + index + '.');
+            }
+        });
     }
 
-    /**
-     * Parse a JSON object for the layers key.
-     *
-     * @since 1.0.0
-     *
-     * @param {object} jsonObj  A parsed JSON object representing the Parchmint file.
-     *
-     * @returns {Array} An array of Layer object containing the data from the layers key in the Parchmint file.
-     */
-    parseLayersArray(jsonObj) {
-        let layers = [];
-        for (let i = 0; i < jsonObj['layers'].length; i++) {
-            let tempLayer = new Layer();
-            this.initParchKey(tempLayer, jsonObj['layers'][i]);
-            layers.push(tempLayer);
+    getParsedLayer(layerObj) {
+        let layerID = layerObj['id'];
+        let tempLayer = new Layer(layerObj['name'], layerID);
+        let tempComps = this.components.get(layerID);
+        let tempConns = this.connections.get(layerID);
+
+        // If a Layer does not have Components or Connections we want to leave their value as the default
+        // empty array.
+        if (tempComps) {
+            tempLayer.components = tempComps;
+        }
+
+        if (tempConns) {
+            tempLayer.connections = tempConns;
```

Comment:
I do like that because it's more concise. I wrote it this way in case the default value is changed in the future, we don't have to change this code. Although, I'm guessing it will most likely never change. What do you think?

Reply:
This is a good point, and I think you are right that we should keep it the way it is.

Reply:
I changed this in favor of expecting ```undefined``` arguments from the ```Map.get``` calls as described by the [comment](https://github.com/brrcrites/printing-press/pull/58#discussion_r313693374) below.